File: iOS/DateMate/Common/Resource.swift
//
//  Resource.swift
//  DateMate
//
//  Created by 홍희표 on 2021/12/16.
//

import Foundation
import Combine

struct Resource<T> {
    var data: T? = nil
    
    var message: String? = nil
    
    var state: State
    
    static func success(data: T) -> Resource {
        return Resource(data: data, state: .Success)
    }
    
    static func error(message: String, data: T? = nil) -> Resource {
        return Resource(data: data, message: message, state: .Error)
    }
    
    static func loading(data: T? = nil) -> Resource {
        return Resource(data: data, state: .Loading)
    }
    
    enum State {
        case Success
        case Error
        case Loading
    }
}

extension Publisher {
    func asResource() -> AnyPublisher<Resource<Output>, Never> {
        return self
            .map(Resource.success)
            .prepend(Resource.loading())
            .catch { error in
                Just(Resource.error(message: error.localizedDescription))
            }
            .eraseToAnyPublisher()
    }
}

File: Android/app/src/main/java/com/hhp227/datemate/common/Resource.kt
package com.hhp227.datemate.common

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.onStart

sealed class Resource<T>(val data: T? = null, val message: String? = null) {
    class Success<T>(data: T) : Resource<T>(data)
    class Error<T>(message: String, data: T? = null) : Resource<T>(data, message)
    class Loading<T>(data: T? = null) : Resource<T>(data)
}

fun <T> Flow<T>.asResource(): Flow<Resource<T>> =
    this
        .map<T, Resource<T>> { Resource.Success(it) }
        .onStart { emit(Resource.Loading()) }
        .catch { e -> emit(Resource.Error(e.message ?: "Unknown error")) }